src/coreclr/nativeaot/Bootstrap/dllmain/dllmain.cpp

```diff
@@ -0,0 +1,16 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <windows.h>
+
+// No-op DllMain. The purpose is to reserve DllMain for our purposes.
+//
+// Some users also try to define a DllMain in managed code using an UnmanagedCallersOnly
+// attribute. This will not work correctly because the entire runtime initialization would run
+// as part of such DllMain, before the first line of users DllMain executes. We don't support
+// initializing the runtime under loader lock.
+//
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+{
```

Comment:
Call `DisableThreadLibraryCalls`?

The weakly-linked default DllMain that we use today does that.

Reply:
Does NativeAOT not static link the CRT in some cases? Static CRT forbids calling `DisableThreadLibraryCalls`.

Reply:
> Does NativeAOT not static link the CRT in some cases?

It does for debug builds only. I guess we can call `DisableThreadLibraryCalls` only when building release bits that link CRT dynamically. Is there a define that can be used to detect whether the compilation targets dynamic or static CRT?

Reply:
> Is there a define that can be used to detect whether the compilation targets dynamic or static CRT?

Seems `#ifndef _DLL` is the documented way for it.

Reply:
Would `DisableThreadLibraryCalls` do anything? The API is documented as "This function does not perform any optimizations if static [Thread Local Storage (TLS)](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage) is enabled. Static TLS is enabled when using thread_local variables, __declspec( thread ) variables, or function-local static."

Reply:
> Would `DisableThreadLibraryCalls` do anything? The API is documented as "This function does not perform any optimizations if static [Thread Local Storage (TLS)](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage) is enabled. Static TLS is enabled when using thread_local variables, __declspec( thread ) variables, or function-local static."

`Do not call this function from a DLL that is linked to the static C run-time library (CRT). The static CRT requires DLL_THREAD_ATTACH and DLL_THREAD_DETATCH notifications to function properly.`

Reply:
> > Would `DisableThreadLibraryCalls` do anything? The API is documented as "This function does not perform any optimizations if static [Thread Local Storage (TLS)](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage) is enabled. Static TLS is enabled when using thread_local variables, __declspec( thread ) variables, or function-local static."
> 
> `Do not call this function from a DLL that is linked to the static C run-time library (CRT). The static CRT requires DLL_THREAD_ATTACH and DLL_THREAD_DETATCH notifications to function properly.`

Is that related?

Reply:
> This function does not perform any optimizations if static [Thread Local Storage (TLS)](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage) is enabled. Static TLS is enabled when using thread_local variables, __declspec( thread ) variables, or function-local static.

Interesting. I did not know that `DisableThreadLibraryCalls` is no-op if the target .dll has thread statics.

Reply:
> > > Would `DisableThreadLibraryCalls` do anything? The API is documented as "This function does not perform any optimizations if static [Thread Local Storage (TLS)](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage) is enabled. Static TLS is enabled when using thread_local variables, __declspec( thread ) variables, or function-local static."
> > 
> > 
> > `Do not call this function from a DLL that is linked to the static C run-time library (CRT). The static CRT requires DLL_THREAD_ATTACH and DLL_THREAD_DETATCH notifications to function properly.`
> 
> Is that related?

That's from the docs of `DisableThreadLibraryCalls`.

Reply:
> Interesting. I did not know that DisableThreadLibraryCalls is no-op if the target .dll has thread statics.

It means that this change is a tiny perf optimization. It will skip the unnecessary `DisableThreadLibraryCalls` call that the boilerplate DllMain does now.